### ISSUE TITLE
dials.stills_process broken for SACLA HDF5 files, possibly due to dxbtx change

### ISSUE THREAD

**Issue by biochem-fan** [dials/dials#725]
I noticed that `dials.stills_process` from the latest DIALS and CCTBX on the GitHub fails to process some (but not all) HDF5 files from SACLA.

An example is `43/Ba-soak/run372134-0.h5` (but not `372134-1` or `372134-2`) from http://cxidb.org/data/43/Ba-soak/ (393 MB).

```
$ dials.stills_process run372134-0.h5
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
Loading files...
Loading run372134-0.h5
Traceback (most recent call last):
  File "/home/nureki/packages/dials/build/../modules/dials/command_line/stills_process.py", line 1323, in <module>
    halraiser(e)
  File "/home/nureki/packages/dials/build/../modules/dials/command_line/stills_process.py", line 1321, in <module>
    script.run()
  File "/home/nureki/packages/dials/build/../modules/dials/command_line/stills_process.py", line 346, in run
    assert len(imageset) == 1
AssertionError: Please report this error to dials-support@lists.sourceforge.net:
```

So I removed the assertion and added the following to `stills_process.py`:
```python
                print(i, len(imageset), imageset)
                #assert len(imageset) == 1
```

The output indicates that `len(imageset)` is 2 for some frames.

```
50 1 <dxtbx_imageset_ext.ImageSet object at 0x7fe6b18ebe68>
51 2 <dxtbx_imageset_ext.ImageSet object at 0x7fe6b18ebec0>
52 1 <dxtbx_imageset_ext.ImageSet object at 0x7fe6b18ebf18>
...
60 1 <dxtbx_imageset_ext.ImageSet object at 0x7fe6b18e7208>
61 1 <dxtbx_imageset_ext.ImageSet object at 0x7fe6b18e7260>
62 2 <dxtbx_imageset_ext.ImageSet object at 0x7fe6b18e72b8>
63 2 <dxtbx_imageset_ext.ImageSet object at 0x7fe6b18e7310>
64 1 <dxtbx_imageset_ext.ImageSet object at 0x7fe6b18e7368>
```

However, the output from `dials.import` is completely fine with all `Experiment`s having an `imageset` with a single image.

This dataset can be processed fine on an older version (`DIALS 1.dev.2676-ga978709`). Copying the `dxtbx` class (`FormatHDF5SaclaMPCCD.py`) from that version did not help. Perhaps this is related to the C++ migration of the `ImageSet`?

**Comment by phyy-nx** [comment]
Hi, I'm downloading the data to investigate.  Thanks.

**Comment by phyy-nx** [comment]
So, this is related to cctbx/cctbx_project#284.  The datablock factory is obviated and in its place the experiment list factory's from_filenames is used.  This returns an experiment list with 68 'crystalless' experiments and 68 imagesets, instead of a datablock with one imageset of length 72.

Check this code out:
```
from dxtbx.model.experiment_list import ExperimentListFactory
from dxtbx.datablock import DataBlockFactory
import sys

filename = sys.argv[1]
datablocks = DataBlockFactory.from_filenames([filename])
print len(datablocks)
datablock = datablocks[0]
print len(datablock.extract_imagesets())
print sum([len(iset) for iset in datablock.extract_imagesets()])

experiments = ExperimentListFactory.from_filenames([filename])
print len(experiments)
print len(experiments.imagesets())
print sum([len(iset) for iset in experiments.imagesets()])
```

Now with this:

```
$ libtbx.python showit.py run372134-0.h5
1
1
72
68
68
72
```

So this file worked before the experiments list changeover, but now is broke.  However, it's because experiment list factory is reading this wrong.

That's as far as I can get tonight.  Tagging @jmp1985.  I'll work on it tomorrow if he doesn't get to it first.

**Comment by graeme-winter** [comment]
@phyy-nx fair to pick this one up over this side of the pond, since it is a side-effect of deprecating datablocks

**Comment by biochem-fan** [comment]
Thank you very much for investigation.

**Comment by graeme-winter** [comment]
🤔 this is interesting

I reproduce @phyy-nx output using the script above, with four experiments having image sets of len > 1 - 

```
for expt in experiments:
    if len(expt.imageset) > 1:
        print expt
```

but then when I do dials.import and then dials.show on the imported_experiments.json file I get 72 experiments:

```
Grey-Area spring8-serial-debug :) $ dials.import run372134-0.h5 
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
DIALS 2.dev.32-g4053b9561
The following parameters have been modified:

input {
  experiments = <image files>
}

--------------------------------------------------------------------------------
  format: <class 'dxtbx.format.FormatHDF5SaclaMPCCD.FormatHDF5SaclaMPCCD'>
  num images: 72
  num sweeps: 0
  num stills: 72
--------------------------------------------------------------------------------
Writing experiments to imported_experiments.json
Grey-Area spring8-serial-debug :) $ dials.show imported_experiments.json | grep Experiment | wc -l
      72

```

**Comment by graeme-winter** [comment]
Also 72 imagesets referenced in this file?

```
Grey-Area spring8-serial-debug :) $ grep imageset imported_experiments.json | awk '{print $NF}'|grep [0-9] | wc -l
      72
```

**Comment by graeme-winter** [comment]
=> dials.import is doing a different thing to ```ExperimentListFactory.from_filenames([filename])```

**Comment by jmp1985** [comment]
@graeme-winter I'm just looking at this issue now

**Comment by graeme-winter** [comment]
👍

**Comment by jmp1985** [comment]
In dxtbx if we had stills and the detector and beam models were the same, then they were part of the same experiment with multiple still images. I've modified this such that stills are always imported into a single image per experiment. Because the input experiments are processed in dials.import to potentially modify geometry, additional checks are done which is why this worked properly in import but not in ExperimentListFactory.from_filenames

**Comment by Anthchirp** [comment]
if you had added "dials/dials#725" to the commit message then people would be able to find that explanation from the commit message ;)

(bonus points for "fixes dials/dials#725")

**Comment by biochem-fan** [comment]
Confirmed on more data files. Thank you very much.
